libraries/lib-dynamic-range-processor/DynamicRangeProcessorUtils.cpp

```diff
@@ -0,0 +1,125 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  DynamicRangeProcessorUtils.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "DynamicRangeProcessorUtils.h"
+#include <optional>
+#include <regex>
+#include <stdexcept>
+#include <unordered_map>
+
+namespace DynamicRangeProcessorUtils
+{
+namespace
+{
+
+template <typename Struct> struct SettingDescription
+{
+   double Struct::*const mem {};
+   const char* const key {};
+};
+
+static const std::vector<SettingDescription<CompressorSettings>>
+   compressorSettingDescriptions {
+      { &CompressorSettings::thresholdDb, "thresholdDb" },
+      { &CompressorSettings::makeupGainDb, "makeupGainDb" },
+      { &CompressorSettings::kneeWidthDb, "kneeWidthDb" },
+      { &CompressorSettings::compressionRatio, "compressionRatio" },
+      { &CompressorSettings::lookaheadMs, "lookaheadMs" },
+      { &CompressorSettings::attackMs, "attackMs" },
+      { &CompressorSettings::releaseMs, "releaseMs" },
+      { &CompressorSettings::showInput, "showInput" },
+      { &CompressorSettings::showOutput, "showOutput" },
+      { &CompressorSettings::showActual, "showActual" },
+      { &CompressorSettings::showTarget, "showTarget" },
+   };
+
+const std::vector<SettingDescription<LimiterSettings>>
+   limiterSettingDescriptions {
+      { &LimiterSettings::thresholdDb, "thresholdDb" },
+      { &LimiterSettings::makeupTargetDb, "makeupTargetDb" },
+      { &LimiterSettings::kneeWidthDb, "kneeWidthDb" },
+      { &LimiterSettings::lookaheadMs, "lookaheadMs" },
+      { &LimiterSettings::releaseMs, "releaseMs" },
+      { &LimiterSettings::showInput, "showInput" },
+      { &LimiterSettings::showOutput, "showOutput" },
+      { &LimiterSettings::showActual, "showActual" },
+      { &LimiterSettings::showTarget, "showTarget" },
+   };
+
+template <typename Struct>
+std::optional<Struct> Deserialize(
+   const std::string& str,
+   const std::vector<SettingDescription<Struct>>& settings)
+{
+   Struct settingStruct;
+
+   const std::regex pattern(R"((\w+)=\"(-?\d+\.?\d*)\")");
+
+   // Create an unordered_map to store the key-value pairs
+   std::unordered_map<std::string, std::string> values;
+
+   const auto begin = std::sregex_iterator(str.begin(), str.end(), pattern);
+   const auto end = std::sregex_iterator();
+
+   for (auto it = begin; it != end; ++it)
+   {
+      const std::smatch match = *it;
+      values[match[1].str()] = match[2].str();
+   }
+
+   if (std::any_of(
+          settings.begin(), settings.end(), [&values](const auto& setting) {
+             return values.find(setting.key) == values.end();
+          }))
+      return {};
+
+   for (const auto& setting : settings)
+      try
+      {
+         settingStruct.*(setting.mem) = std::stod(values.at(setting.key));
```

Comment:
std::stod can throw out_of_range exception which is not handled

Reply:
Should be caught by the unit tests, but you're right, I'll generalize.